contracts/command-batcher/src/encoding/evm.rs

```diff
@@ -0,0 +1,557 @@
+
+    Keccak256::digest(&pub_key.as_bytes()[1..]).as_slice()[12..].into()
+}
+
+impl traits::Proof for Proof {
+    fn new(
+        snapshot: Snapshot,
+        signers: HashMap<String, Signature>,
+        pub_keys: HashMap<String, multisig::types::PublicKey>,
+    ) -> Self {
+        let mut operators = snapshot
+            .participants
+            .iter()
+            .map(|(_, participant)| {
+                let pub_key = pub_keys
+                    .get(&participant.address.to_string())
+                    .expect("violated invariant: participant address not found in pub_keys")
+                    .into();
+
+                Operator {
+                    address: evm_address(pub_key),
+                    weight: participant.weight.into(),
+                    signature: signers.get(participant.address.as_str()).cloned(),
+                }
+            })
+            .collect::<Vec<Operator>>();
+        operators.sort_by(|a, b| a.address.cmp(&b.address));
+
+        Proof {
+            operators,
+            threshold: snapshot.quorum.into(),
+        }
+    }
+
+    fn encode_execute_data(&self, data: &Data) -> HexBinary {
+        let data_encoded = data.encode();
+        let proof_encoded = self.encode();
+
+        ethabi::encode(&[
+            Token::Bytes(data_encoded.into()),
+            Token::Bytes(proof_encoded.into()),
+        ])
+        .into()
+    }
+}
+
+impl traits::CommandBatch for CommandBatch {
+    fn new(block_height: u64, messages: Vec<Message>, destination_chain_id: Uint256) -> Self {
+        let message_ids = messages.iter().map(|msg| msg.id.clone()).collect();
+
+        let data = Data::new(destination_chain_id, messages);
+        let encoded_data = data.encode();
+
+        let id = batch_id(block_height, &encoded_data);
+        let msg_to_sign = msg_to_sign(&encoded_data);
+
+        Self {
+            id,
+            message_ids,
+            data,
+            msg_to_sign,
+            multisig_session_id: None,
+        }
+    }
+}
+
+fn command_id(message_id: String) -> [u8; 32] {
+    // TODO: we might need to change the command id format to match the one in core for migration purposes
+
+    Keccak256::digest(message_id.as_bytes()).into()
+}
+
+// TODO: This will make it incompatible with current version of destination chain gateways,
+// they rely on tx hash and event index as well, but just to emit events
+// https://github.com/axelarnetwork/axelar-cgp-solidity/blob/main/contracts/AxelarGateway.sol#L466
+fn command_params(
+    source_chain: String,
+    source_address: String,
+    destination_address: ethereum_types::Address,
+    payload_hash: [u8; 32],
+) -> HexBinary {
+    ethabi::encode(&[
+        Token::String(source_chain),
+        Token::String(source_address),
+        Token::Address(destination_address),
+        Token::FixedBytes(payload_hash.into()),
+    ])
+    .into()
+}
+
+fn batch_id(block_height: u64, data: &HexBinary) -> HexBinary {
```

Comment:
What do we need the block height?

Reply:
honestly I don't remember, but this is how core is creating the ID right now. I tried to make the output as similar to core as possible. Maybe @sammy1991106 can enlighten us?

Reply:
Unless there's a good reason, I'd rather remove it because I feel like it makes things confusing.